build(client): generate eslint 9 configs for all packages

tylerbutler opened the pull request:
This change updates the shared eslint 9 config to be TypeScript and updates the script in scripts/generate-flat-eslint-configs.ts to support a `--typescript` flag to output typescript per-package configs instead of ESM. Also adds a `--finalize` flag that will do some additional cleanup once packages are fully migrated to eslint 9.

Finally, there are lots of fixes to the script to handle ignore entries, inheritance from other configs, etc. The script now produces configs that work in my testing without any modifications. Even if that doesn't hold true once we are truly migrating (see #25932), merging this beforehand will simplify the review of those changes.

All the configs were generated by the script using this command:

```
tsx scripts/generate-flat-eslint-configs.ts --typescript
```

tylerbutler committed:
chore: rename flat.mjs to flat.mts

tylerbutler committed:
use typescript for flat config

tylerbutler committed:
Merge branch 'main' into eslint-rename-configs

tylerbutler committed:
update script and config

tylerbutler committed:
format

tylerbutler committed:
generate eslint 9 configs for all packages

tylerbutler committed:
better handling of config inheritance

tylerbutler committed:
Merge branch 'main' into eslint-rename-configs

alexvy86 reviewed (commented):
Only looked at a few of the generated configs; didn't see anything particularly weird, but it was far from an exhaustive review. I still feel ok overall moving forward with this.

The comment for the config.data.mts file I think does need addressing.

alexvy86 commented on common/build/eslint-config-fluid/flat.mts:
```diff
@@ -0,0 +1,328 @@
+// Global ignores for all configs.
+// In flat config, a config object with only `ignores` is treated as a global ignore pattern.
+// See: https://eslint.org/docs/latest/use/configure/configuration-files#globally-ignoring-files-with-ignores
+const globalIgnores = {
```
Nit: Eslint docs suggest using the `globalIgnores()` helper for this.

Using `defineConfig()` like most of the Eslint examples would also be nice but seem more complicated here, and I'm not sure if it gives us any extra value if we're already using the correct types (e.g. `FlatConfigArray`) when defining our configs.

tylerbutler replied:
My plan is to completely rewrite these as native flat configs once we are on 9. So I'll make a note of this for when we do that part.

tylerbutler committed:
Merge branch 'main' into eslint-rename-configs

tylerbutler committed:
fix(eslint): export named lists from shared eslint config

Update the generate-flat-eslint-configs script to preserve named exports
from .eslintrc.data.cjs files when generating flat configs. This allows
child packages to import and extend the individual lists (e.g.,
importInternalModulesAllowed) rather than just the composed config.

The eslint.config.data.mts now exports:
- importInternalModulesAllowed: base list of allowed imports
- importInternalModulesAllowedForTest: extended list for test files
- default: the composed config array

This addresses PR feedback about maintaining the original purpose of
the shared config file.

tylerbutler committed:
fix(eslint): use named exports from shared config in child packages

Update external-data and webflow eslint configs to import and use
the named exports (importInternalModulesAllowed, importInternalModulesAllowedForTest)
from eslint.config.data.mts instead of inlining the expanded arrays.

This allows these packages to properly extend the base lists while
adding their own custom patterns, matching the original behavior
of the legacy .eslintrc.cjs configs.

tylerbutler committed:
format

tylerbutler committed:
rm unused files

alexvy86 reviewed (approved):
I looked at some of the generated configs, and skimmed the script to generate them. No major concerns. Since this has no effect yet and we'll almost certainly know if there are any issues when we swap the packages to Eslint9, I think this is fine to merge and we can fix issues as we discover them.
